custom_components/zaptec/api.py

```diff
@@ -399,327 +737,265 @@ async def _refresh_token(self):
+                _LOGGER.debug(f"@@@  RESPONSE {resp.status} length {len(contents)}")
+                _LOGGER.debug(
+                    f"     header {dict((k, v) for k, v in resp.headers.items())}"
+                )
+                if not contents:
+                    return
+                if resp.status != 200:
+                    _LOGGER.debug(f"     content {await resp.text()}")
+                else:
+                    _LOGGER.debug(f"     json '{await resp.json(content_type=None)}'")
+            except Exception as err:
+                _LOGGER.exception("Failed to log response")
 
-    async def _request(self, url, method="get", data=None):
         header = {
-            "Authorization": "Bearer %s" % self._access_token,
+            "Authorization": f"Bearer {self._access_token}",
             "Accept": "application/json",
         }
         full_url = API_URL + url
         try:
-            with async_timeout.timeout(30):
+            async with async_timeout.timeout(API_TIMEOUT):
+                if DEBUG_API_CALLS:
+                    log_request()
+
                 call = getattr(self._client, method)
-                if data is not None and method == "post":
+                if data is not None and method in ("post", "put"):
                     call = partial(call, json=data)
+
+                resp: aiohttp.ClientResponse
                 async with call(full_url, headers=header) as resp:
+                    if DEBUG_API_CALLS:
+                        await log_response(resp)
+
                     if resp.status == 401:  # Unauthorized
                         await self._refresh_token()
-                        return await self._request(url)
+                        if iteration > API_RETRIES:
+                            raise RequestRetryError(
+                                f"Request to {full_url} failed after {iteration} retries"
+                            )
+                        return await self._request(url, iteration=iteration + 1)
+
                     elif resp.status == 204:  # No content
                         content = await resp.read()
                         return content
+
                     elif resp.status == 200:  # OK
+                        # FIXME: This will raise json error if the json is invalid. How to handle this?
                         json_result = await resp.json(content_type=None)
-                        return json_result
-                    else:
-                        _LOGGER.error("Could not get info from %s: %s", full_url, resp)
-                        return None
 
-        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
-            _LOGGER.error("Could not get info from %s: %s", full_url, err)
+                        # Validate the incoming json data
+                        # FIXME: This raise pydantic.ValidationError if the json is unexpected. How to handle this?
+                        validate(json_result, url=url)
```

Comment:
Why validate?

Reply:
Because we cannot assume that the data from the Zaptec API is formatted correctly, one should always validate incoming data. What if Zaptec changes something?

We rely on certain data being present, e.g. `for data in installations["Data"]:` assumes `Data` is there. If it isn't then the resulting error can be very hard to link to the incoming data being malformed. Validation ensures that one gets a clear "I don't understand the data I just got"-type errors, which is a lot easier to deal with.

There is normally another dimension to validation and that is deserialization into named objects (see validation.py). But given the way this API is constructed with the special attribute handling, this feature isn't used.